addons/website_livechat/models/website_visitor.py

```diff
@@ -33,6 +36,38 @@ def _compute_livechat_operator_id(self):
         for visitor in self:
             visitor.livechat_operator_id = visitor_operator_map.get(visitor.id, False)
 
+    @api.depends("last_connection_datetime")
+    def _compute_page_visit_history(self):
+        if not self:
+            return
+        query = """
+                    SELECT visitor_id, page_name, visit_datetime
+                      FROM (
+                          SELECT website_track.visitor_id,
+                                 ir_ui_view.name as page_name,
+                                 website_track.visit_datetime,
+                                 ROW_NUMBER() OVER(
+                                    PARTITION BY website_track.visitor_id ORDER BY website_track.visit_datetime DESC
+                                 ) as row_number
+                            FROM website_track
+                            JOIN website_page ON website_page.id = website_track.page_id
+                            JOIN ir_ui_view ON ir_ui_view.id = website_page.view_id
+                           WHERE website_track.visitor_id IN %s
+                             AND website_track.page_id IS NOT NULL
+                      ) as ranked_visits
+                     WHERE row_number <= 3
+                  ORDER BY visitor_id, visit_datetime ASC
+                """
+        self.env.cr.execute(query, (tuple(self.ids),))
+        results = self.env.cr.dictfetchall()
+        history_map = defaultdict(list)
+        for res in results:
+            history_map[res["visitor_id"]].append(
+                (res["page_name"], fields.Datetime.to_string(res["visit_datetime"]))
+            )
+        for visitor in self:
+            visitor.page_visit_history = history_map.get(visitor.id, [])
```

Comment:
It feels strange to send a json field, it looks like we should be having the model for website.track and sending that as a relation directly instead.